src/transport/MessageEndpoint.cpp

```diff
@@ -0,0 +1,119 @@
+#include <faabric/transport/MessageEndpoint.h>
+#include <faabric/util/locks.h>
+
+#include <faabric/util/gids.h>
+#include <faabric/util/logging.h>
+
+namespace faabric::transport {
+MessageEndpoint::MessageEndpoint(const std::string& hostIn, int portIn)
+  : host(hostIn)
+  , port(portIn)
+  , id(faabric::util::generateGid())
+{}
+
+MessageEndpoint::~MessageEndpoint()
+{
+    if (this->socket) {
+        this->close();
+    }
+}
+
+void MessageEndpoint::open(faabric::transport::MessageContext& context,
+                           faabric::transport::SocketType sockType,
+                           bool bind)
+{
+    faabric::util::getLogger()->info("Opening socket: {}", id);
+    std::string address =
+      "tcp://" + this->host + ":" + std::to_string(this->port);
+
+    // Note - only one socket may bind, but several can connect. This allows
+    // for easy N - 1 or 1 - N PUSH/PULL patterns. Order between bind and
+    // connect does not matter.
+    switch (sockType) {
+        case faabric::transport::SocketType::PUSH:
+            this->socket = std::make_unique<zmq::socket_t>(
+              context.get(), zmq::socket_type::push);
+            break;
+        case faabric::transport::SocketType::PULL:
+            this->socket = std::make_unique<zmq::socket_t>(
+              context.get(), zmq::socket_type::pull);
+            break;
+        default:
+            throw std::runtime_error("Unrecognized socket type");
+    }
+
+    // Set linger period to zero. This avoids blocking if closing a socket with
+    // outstanding messages.
+    this->socket->set(zmq::sockopt::linger, 0);
+
+    // Bind or connect the socket
+    if (bind) {
+        this->socket->bind(address);
+    } else {
+        this->socket->connect(address);
+    }
+}
+
+void MessageEndpoint::send(char* serialisedMsg, size_t msgSize, bool more)
+{
+    if (!this->socket) {
+        throw std::runtime_error("Trying to send from a null-pointing socket");
+    }
+
+    // Pass a deallocation function for ZeroMQ to be zero-copy
+    zmq::message_t msg(serialisedMsg, msgSize, [](void* data, void* hint) {
+        delete[](char*) data;
+    });
+
+    if (more) {
+        if (!this->socket->send(msg, zmq::send_flags::sndmore)) {
+            throw std::runtime_error("Error sending message through socket");
+        }
+    } else {
+        if (!this->socket->send(msg, zmq::send_flags::none)) {
+            throw std::runtime_error("Error sending message through socket");
+        }
+    }
+}
+
+// This method overload is used to send flatbuffers that deal with unsigned
+// chars, and are stack-allocated.
+void MessageEndpoint::send(uint8_t* serialisedMsg, size_t msgSize, bool more)
```

Comment:
The alternative was to define a template function, and check the argument type with something like:
```cpp
if constexpr(std::is_same_v<T, char>)
```
but in the end it added more complexity imo.

Reply:
Having the `char*` version delete the data that's being passed in is quite unexpected IMO. If I'm passing data to a function via a pointer, I wouldn't expected it to be deleted. Do we really need to pass 0MQ a deallocation function to avoid it doing a copy? Unless you can't avoid it, I think it's safest to assume the caller is responsible for deleting a pointer to data they have created.

Also, couldn't we change the `char` version to be:

```
void MessageEndpoint::send(char* serialisedMsg, size_t msgSize, bool more)
{
    send(reinterpret_cast<uint8_t *>(serialisedMsg), msgSize, more);    
}
```